doc/man/man1/rsplit.1
.\" RCSid "$Id: rsplit.1,v 1.13 2024/05/15 18:58:54 greg Exp $"
.TH RSPLIT 1 7/8/97 RADIANCE
.SH NAME
rsplit - splits standard input into multiple output files or streams
.SH SYNOPSIS
.B rsplit
[
.B \-tS
][
.B \-ih
][
.B \-iH
][
.B \-oh
][
.B \-oH
][
.B "\-x xres \-y yres"
][
.B \-oaN
|
.B \-ofN
|
.B \-odN
|
.B \-oiN
|
.B \-owN
|
.B \-obN
][
.B \-f
][
.B \-a
][
.B "\-on M"
]
output1 [options] output2 ..
.SH DESCRIPTION
.I Rsplit
writes lines or fields from the standard input to one or more output
streams.
By default, fields are separated by a
newline character (i.e., input is distributed by lines),
but the terminator can be changed using the
.I \-t
option.
If a space is specified as the terminator, then fields are separated by
any number of white spaces, including tabs, returns, and newlines.
Different terminators may be given for different fields by specifying
additional
.I \-t
options between each output specification.
Note that there is no space between this option and its argument, and
certain arguments, such as a space or a tab, must be quoted or escaped
with a backslash in most shells.
If no argument is given, the default newline terminator is re-established.
.PP
Output is either to a file, to a command, to the standard output, or to discard.
If a named output file exists, it will not be overwritten unless the
.I \-f
option is given, or it is being appended using the
.I \-a
option.
The same output file may not appear more than once.
Commands are given in quotes, and begin with an exclamantion point ('!').
At most, one command may reliably write to its standard output, and
doing so is incompatible with
.I rsplit
also writing to standard output.
A hyphen ('-') by itself tells
.I rsplit
to write directly to its standard output, and may appear multiple times.
The options before the first appearance determine the
standard output header parameters.
A period ('.') tells
.I rsplit
to discard the corresponding data, and also may appear multiple times.
.PP
The
.I \-oa
option may be used to specify ASCII data (the default), or the
.I \-of
option may be used to indicated binary IEEE 32-bit floats.
Similarly, the
.I \-od
and
.I \-oi
options may be used to indicate binary 64-bit doubles or integer words,
respectively.
The
.I \-ow
option specifies 2-byte short words, and the
.I \-ob
option specifies bytes.
If a number immediately follows any of these options, then it
indicates that multiple such values are expected for each record.
For example,
.I \-of3
indicates three floats per output record for the next named output.
In the case of the
.I \-oa
option, no number writes one line or field per stream, and numbers
greater than zero write multiple fields per record to the same stream.
If the terminating character is set to something other than newline ('\\n'),
a newline is added after each record (in addition to the terminator).
For binary output formts, no number implies one value per output.
No terminator characters are expected or written for binary outputs.
.PP
If a
.I \-on
option is given with a positive integer argument, this will be the
maximum number of records that will be written by
.I rsplit.
.PP
The
.I \-ih
option tells
.I rsplit
to expect a Radiance header on its input.
The
.I \-iH
option says to expect a resolution string for an image or similar.
The
.I \-x
and
.I \-y
options may alternatively be used to specify the input dimensions.
The
.I \-oh
option tells
.I rsplit
to produce a Radiance header on the next output stream.
The
.I \-oH
option will write a resolution string.
These options toggle the associated mode on and off,
so it is possible to have one output include a header or
resolution string and not the next by specifying the option(s) again.
Neither
.I \-oh
nor
.I \-oH
may be used with the append mode ('-a'), which may also be toggled.
.SH EXAMPLES
To alternate 5 lines between two output files:
.IP "" .2i
rsplit -oa5 output1.txt output2.txt < input.txt
.PP
To send the first of each double precision triplet to one file, and the second
and third to a second file, stopping after 500 triplets:
.IP "" .2i
rsplit -on 500 -od firstvals.dbl -od2 followingpairs.dbl < triplets.dbl
.PP
To convey material, color and distance information from rtrace to
separate files:
.IP "" .2i
vwrays [viewspec] | rtrace [options] -x 1024 -y 768 -h+ -oMlv octree | rsplit -ih -iH -t^I mats.txt -of depth.zbf -oh -oH -of3 '!pvalue -r -df > image.hdr'
.SH AUTHOR
Greg Ward
.SH "SEE ALSO"
cnt(1), histo(1), neaten(1), rcalc(1), rcode_depth(1),
rcode_ident(1), rcode_norm(1), rcode2bmp(1), rcrop(1),
rlam(1), rtpict(1), rtrace(1), split(1), tabfunc(1), total(1)


src/cal/rsplit.c
#ifndef lint
static const char	RCSid[] = "$Id: rsplit.c,v 1.15 2020/04/05 15:07:09 greg Exp $";
#endif
/*
 *  rsplit.c - split input into multiple output streams
 *
 *	7/4/19		Greg Ward
 */

#include <ctype.h>

#include "rtio.h"
#include "platform.h"
#include "paths.h"
#include "resolu.h"

#define	DOHEADER	1
#define DORESOLU	2

#define MAXFILE		512		/* maximum number of files */

static int		swapped = 0;	/* input is byte-swapped */

static FILE		*output[MAXFILE];
static int		ncomp[MAXFILE];
static int		bytsiz[MAXFILE];
static int		hdrflags[MAXFILE];
static const char	*format[MAXFILE];
static int		termc[MAXFILE];
static int		nfiles = 0;

static RESOLU	ourres = {PIXSTANDARD, 0, 0};

static char	buf[16384];		/* input buffer used in scanOK() */


/* process header line */
static int
headline(char *s, void *p)
{
	extern const char	FMTSTR[];
	int			i;

	if (strstr(s, FMTSTR) == s)
		return(0);		/* don't copy format */
	if (!strncmp(s, "NROWS=", 6))
		return(0);
	if (!strncmp(s, "NCOLS=", 6))
		return(0);
	if (!strncmp(s, "NCOMP=", 6))
		return(0);
	if ((i = isbigendian(s)) >= 0) {
		swapped = (nativebigendian() != i);
		return(0);
	}
	i = nfiles;
	while (i--)			/* else copy line to output streams */
		if (hdrflags[i] & DOHEADER)
			fputs(s, output[i]);
	return(1);
}


/* scan field into buffer up to and including terminating byte */
static int
scanOK(int termc)
{
	int	skip_white = (termc == ' ');
	char	*cp = buf;
	int	c;

	while ((c = getchar()) != EOF) {
		if (skip_white && isspace(c))
			continue;
		skip_white = 0;
		if (c == '\n' && isspace(termc))
			c = termc;	/* forgiving assumption */
		*cp++ = c;
		if (cp-buf >= sizeof(buf))
			break;
		if ((termc == ' ') ? isspace(c) : (c == termc)) {
			*cp = '\0';
			return(cp-buf);
		}
	}
	return(0);
}


int
main(int argc, char *argv[])
{
	int		inpflags = 0;
	int		needres = 0;
	int		force = 0;
	int		append = 0;
	long		outcnt = 0;
	int		bininp = 0;
	int		nstdoutcomp = 0;
	int		curterm = '\n';
	int		curncomp = 1;
	int		curbytes = 0;
	int		curflags = 0;
	const char	*curfmt = "ascii";
	short		outndx[MAXFILE];
	int		i;

	memset(outndx, 0, sizeof(outndx));
	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {
			case 't':
				curterm = argv[i][2];
				if (!curterm) curterm = '\n';
				break;
			case 'i':
				switch (argv[i][2]) {
				case 'h':
					inpflags ^= DOHEADER;
					break;
				case 'H':
					inpflags ^= DORESOLU;
					break;
				default:
					goto badopt;
				}
				break;
			case 'f':
				force = !force;
				break;
			case 'a':
				append = !append;
				break;
			case 'x':
				ourres.xr = atoi(argv[++i]);
				break;
			case 'y':
				ourres.yr = atoi(argv[++i]);
				break;
			case 'o':
				switch (argv[i][2]) {
				case 'n':
					outcnt = atol(argv[++i]);
					continue;
				case 'h':
					curflags ^= DOHEADER;
					continue;
				case 'H':
					curflags ^= DORESOLU;
					continue;
				case 'f':
					curfmt = "float";
					curbytes = sizeof(float);
					break;
				case 'd':
					curfmt = "double";
					curbytes = sizeof(double);
					break;
				case 'i':
					curfmt = "int";
					curbytes = sizeof(int);
					break;
				case 'w':
					curfmt = "16-bit";
					curbytes = 2;
					break;
				case 'b':
					curfmt = "byte";
					curbytes = 1;
					break;
				case 'a':
					curfmt = "ascii";
					curbytes = 0;
					break;
				default:
					goto badopt;
				}
				curncomp = isdigit(argv[i][3]) ?
							atoi(argv[i]+3) : 1 ;
				if (curbytes*curncomp > (int)sizeof(buf)) {
					fputs(argv[0], stderr);
					fputs(": output size too big\n", stderr);
					return(1);
				}
				bininp += (curbytes > 0);
				break;
			case '\0':			/* stdout */
				if (!nstdoutcomp) {	/* first use? */
					needres |= (curflags & DORESOLU);
					hdrflags[nfiles] = curflags;
				}
				output[nfiles] = stdout;
				if (curbytes > 0)
					SET_FILE_BINARY(output[nfiles]);
				termc[nfiles] = curterm;
				format[nfiles] = curfmt;
				nstdoutcomp +=
					ncomp[nfiles] = curncomp;
				bytsiz[nfiles] = curbytes;
				outndx[nfiles++] = i;
				break;
			badopt:;
			default:
				fputs(argv[0], stderr);
				fputs(": bad option\n", stderr);
				return(1);
			}
		} else if (argv[i][0] == '.' && !argv[i][1]) {
			output[nfiles] = NULL;		/* discard data */
			termc[nfiles] = curterm;
			format[nfiles] = curfmt;
			ncomp[nfiles] = curncomp;
			bytsiz[nfiles] = curbytes;
			outndx[nfiles++] = i;
		} else if (argv[i][0] == '!') {
			needres |= (curflags & DORESOLU);
			hdrflags[nfiles] = curflags;
			termc[nfiles] = curterm;
			if ((output[nfiles] = popen(argv[i]+1, "w")) == NULL) {
				fputs(argv[i], stderr);
				fputs(": cannot start command\n", stderr);
				return(1);
			}
			if (curbytes > 0)
				SET_FILE_BINARY(output[nfiles]);
			format[nfiles] = curfmt;
			ncomp[nfiles] = curncomp;
			bytsiz[nfiles] = curbytes;
			outndx[nfiles++] = i;
		} else {
			int	j = nfiles;
			while (j--)			/* check duplicates */
				if (!strcmp(argv[i], argv[outndx[j]])) {
					fputs(argv[0], stderr);
					fputs(": duplicate output: ", stderr);
					fputs(argv[i], stderr);
					fputc('\n', stderr);
					return(1);
				}
			if (append & (curflags != 0)) {
				fputs(argv[0], stderr);
				fputs(": -a option incompatible with -oh and -oH\n",
						stderr);
				return(1);
			}
			needres |= (curflags & DORESOLU);
			hdrflags[nfiles] = curflags;
			termc[nfiles] = curterm;
			if (!append & !force && access(argv[i], F_OK) == 0) {
				fputs(argv[i], stderr);
				fputs(": file exists -- use -f to overwrite\n",
						stderr);
				return(1);
			}
			output[nfiles] = fopen(argv[i], append ? "a" : "w");
			if (output[nfiles] == NULL) {
				fputs(argv[i], stderr);
				fputs(": cannot open for output\n", stderr);
				return(1);
			}
			if (curbytes > 0)
				SET_FILE_BINARY(output[nfiles]);
			format[nfiles] = curfmt;
			ncomp[nfiles] = curncomp;
			bytsiz[nfiles] = curbytes;
			outndx[nfiles++] = i;
		}
		if (nfiles >= MAXFILE) {
			fputs(argv[0], stderr);
			fputs(": too many output streams\n", stderr);
			return(1);
		}
	}
	if (!nfiles) {
		fputs(argv[0], stderr);
		fputs(": no output streams\n", stderr);
		return(1);
	}
	if (bininp)				/* binary input? */
		SET_FILE_BINARY(stdin);
#ifdef getc_unlocked				/* avoid lock/unlock overhead */
	flockfile(stdin);
	for (i = nfiles; i--; )
		if (output[i] != NULL)
			ftrylockfile(output[i]);
#endif
						/* load/copy header */
	if (inpflags & DOHEADER && getheader(stdin, headline, NULL) < 0) {
		fputs(argv[0], stderr);
		fputs(": cannot read header from standard input\n",
				stderr);
		return(1);
	}
						/* handle resolution string */
	if (inpflags & DORESOLU && !fgetsresolu(&ourres, stdin)) {
		fputs(argv[0], stderr);
		fputs(": bad resolution string on standard input\n", stderr);
		return(1);
	}
	if (needres && (ourres.xr <= 0) | (ourres.yr <= 0)) {
		fputs(argv[0], stderr);
		fputs(": -oH option requires -iH or -x and -y options\n", stderr);
		return(1);
	}
	if ((ourres.xr > 0) & (ourres.yr > 0)) {
		if (outcnt <= 0) {
			outcnt = ourres.xr * ourres.yr;
		} else if (outcnt != ourres.xr*ourres.yr) {
			fputs(argv[0], stderr);
			fputs(": warning: -on option does not agree with resolution\n",
					stderr);
		}
	}
	for (i = 0; i < nfiles; i++) {		/* complete headers */
		if (hdrflags[i] & DOHEADER) {
			if (!(inpflags & DOHEADER))
				newheader("RADIANCE", output[i]);
			printargs(argc, argv, output[i]);
			fprintf(output[i], "NCOMP=%d\n", output[i]==stdout ?
						nstdoutcomp : ncomp[i]);
			if (format[i] != NULL) {
				extern const char  BIGEND[];
				if (bytsiz[i] > 1) {
					fputs(BIGEND, output[i]);
					fputs(nativebigendian() ^ swapped ?
						"1\n" : "0\n", output[i]);
				}
				fputformat(format[i], output[i]);
			}
			fputc('\n', output[i]);
		}
		if (hdrflags[i] & DORESOLU)
			fputsresolu(&ourres, output[i]);
	}
	do {					/* main loop */
		for (i = 0; i < nfiles; i++) {
			if (bytsiz[i] > 0) {		/* binary output */
				if (getbinary(buf, bytsiz[i], ncomp[i],
							stdin) < ncomp[i])
					break;
				if (output[i] != NULL &&
					    putbinary(buf, bytsiz[i], ncomp[i],
							output[i]) < ncomp[i])
					break;
			} else if (ncomp[i] > 1) {	/* N-field output */
				int	n = ncomp[i];
				while (n--) {
					if (!scanOK(termc[i]))
						break;
					if (output[i] != NULL &&
						    fputs(buf, output[i]) == EOF)
						break;
				}
				if (n >= 0)		/* fell short? */
					break;
				if ((output[i] != NULL) &  /* add EOL if none */
						(termc[i] != '\n'))
					fputc('\n', output[i]);
			} else {			/* 1-field output */
				if (!scanOK(termc[i]))
					break;
				if (output[i] != NULL) {
					if (fputs(buf, output[i]) == EOF)
						break;
					if (termc[i] != '\n')	/* add EOL? */
						fputc('\n', output[i]);
				}
			}
							/* skip input EOL? */
			if (!bininp && termc[nfiles-1] != '\n') {
				int	c = getchar();
				if ((c != '\n') & (c != EOF))
					ungetc(c, stdin);
			}
		}
		if (i < nfiles)
			break;
	} while (--outcnt);
							/* check ending */
	if (fflush(NULL) == EOF) {
		fputs(argv[0], stderr);
		fputs(": write error on one or more outputs\n", stderr);
		return(1);
	}
	if (outcnt > 0) {
		fputs(argv[0], stderr);
		fputs(": warning: premature EOD\n", stderr);
	}
	return(0);
}
